### ISSUE TITLE
Publishing fails when using term with keyref and chunking

### ISSUE THREAD

**Issue by raducoravu** [dita-ot/dita-ot#2243]
I tested using the latest DITA OT from the 2.2.3 branch on Windows, uploaded samples here:

http://www.oxygenxml.com/forum/files/chunkProblemWithTerm.zip

The failure is something like:

         [xslt] Processing C:\Users\radu_coravu\Desktop\zuzu\temp\xhtml\topics\register-license-key.dita to C:\Users\radu_coravu\Desktop\zuzu\out\xhtml\topics\register-license-key.html
         [xslt] C:\Users\radu_coravu\Documents\dot-develop\src\main\xsl\common\dita-utilities.xsl:394: Fatal Error! Base URI {file:/C:/Users/radu_coravu/Des...} is invalid: Invalid relative URI: Illegal character in path at index 2: ..\glossary\named-user.dita: file:/C:/Users/radu_coravu/Desktop/zuzu/temp/xhtml/topics/register-license-key.dita
         [xslt] Failed to process null
            
            BUILD FAILED
            C:\Users\radu_coravu\Documents\dot-develop\src\main\build.xml:41: The following error occurred while executing this line:
            C:\Users\radu_coravu\Documents\dot-develop\src\main\plugins\org.dita.xhtml\build_general.xml:162: Fatal error during transformation using C:\Users\radu_coravu\Documents\dot-develop\src\main\plugins\org.dita.xhtml\xsl\dita2xhtml.xsl: Base URI {file:/C:/Users/radu_coravu/Des...} is invalid: Invalid relative URI: Illegal character in path at index 2: ..\glossary\named-user.dita: file:/C:/Users/radu_coravu/Desktop/zuzu/temp/xhtml/topics/register-license-key.dita; SystemID: file:/C:/Users/radu_coravu/Documents/dot-develop/src/main/xsl/common/dita-utilities.xsl; Line#: 394; Column#: -1

**Comment by raducoravu** [comment]
Looking at how the "register-license-key.dita" looks like in the temporary files folder, it has a link like this:

      <term keyref="glossentry_named-user" class="- topic/term " xtrf="file:/C:/Users/radu_coravu/Desktop/zuzu/tasks/unregister-license-key.dita" xtrc="term:1;42:97" href="..\glossary\named-user.dita" linking="none"/>

Because that "\" is used in the @href, the resolve-uri method breaks in the XSLT.

**Comment by raducoravu** [comment]
I think the entire DITA OT Java code should be combed for places where the Windows file separator is explicitly used or the File.separator is used, and all those replaced with the Linux style separator "/" which works both as well on Windows.

**Comment by raducoravu** [comment]
This worked with the first DITA OT 2.x versions. At some point @eerohele added this XSLT code:
https://github.com/eerohele/dita-ot/commit/d478cc709833311207d17d7eff76dbc0451c02b7

and this indirectly triggered the problem by passing the reference through the resolve-uri XSLT method.

**Comment by eerohele** [comment]
Yes, I introduced this regression when fixing #1816. I didn't think to test the change with Windows-style separators.

I actually also thought that DITA-OT already converted Windows separators to Unix separators. Turns out I was wrong.

**Comment by raducoravu** [comment]
Yes, the problem is deeper embedded in the DITA OT. Maybe a simple fix could be made for this situation but ideally we could avoid using the Windows separator in the entire DITA OT code. Should I add a new issue for the more general refactor task?

**Comment by eerohele** [comment]
Just as a side note: according to [the spec](https://www.oxygenxml.com/dita/1.3/specs/#langRef/attributes/thehrefattribute.html), the value of `@href` needs to be a valid URI, which means that `\` is not a permitted character:

>The value of a DITA `@href` attribute must be a valid URI reference [RFC 3986]. It is an error if the value is not a valid URI reference. An implementation MAY generate an error message; it MAY recover from this error condition by attempting to convert the value to a valid URI reference. Note that the path separator character in a URI is the forward slash (“/”); the backward slash character (“\”) is not permitted unescaped within URIs.

So DITA-OT doesn't *need* to recover from this error (although even in that case a more helpful error message would be good), but it would certainly increase usability to try to convert `@href` values to valid URIs.

A quick fix would be to do it in the `dita-ot:resolve-href-path` function.

**Comment by jelovirt** [comment]
Not sure if it would change anything, but https://github.com/dita-ot/dita-ot/blob/develop/src/main/java/org/dita/dost/writer/ValidationFilter.java#L216 should fix all URI attributes, e.g. `@conref` and `object/@data` too.

**Comment by raducoravu** [comment]
The original content has no "\" inside it. The problem is that after some preprocessing one of the files in the temporary files folder has an "\" in its @href which was automatically generated (because it is a term with @keyref).

**Comment by jelovirt** [comment]
@raducoravu Yep, clearly a regression bug then.

**Comment by eerohele** [comment]
@raducoravu Ah, I see — I didn't look at the input files yet. Yeah, then it's definitely a regression (although I'd say this could be considered a regression even in cases where the authored content had a backslash).

**Comment by jelovirt** [comment]
The first task would be to track which processing step creates the invalid URI. Unfortunately I don't have access to a Windows machine, so looking for volunteers. I can assist in debugging if help is needed.

**Comment by eerohele** [comment]
@jelovirt: I can do that.

**Comment by eerohele** [comment]
@jelovirt: I don't have time to dive into the `URLUtils` class to find what the "idiomatic" way to fix this is at the moment, but this shows where the problem originates and one way to fix it:

```diff
diff --git a/src/main/java/org/dita/dost/writer/TopicRefWriter.java b/src/main/java/org/dita/dost/writer/TopicRefWriter.java
index b466f01..e576b4d 100644
--- a/src/main/java/org/dita/dost/writer/TopicRefWriter.java
+++ b/src/main/java/org/dita/dost/writer/TopicRefWriter.java
@@ -11,6 +11,7 @@ package org.dita.dost.writer;
 import static org.dita.dost.util.Constants.*;
 import static org.dita.dost.util.FileUtils.*;
 import static org.dita.dost.util.XMLUtils.*;
+import static org.dita.dost.util.URLUtils.toURI;
 
 import java.io.File;
 import java.util.Map;
@@ -178,7 +179,7 @@ public final class TopicRefWriter extends AbstractXMLFilter {
             }
 
             if (changeTarget == null) {
-                return hrefValue;// no change
+                return toURI(hrefValue).toASCIIString();
             } else {
                 final String conTarget = conflictTable.get(stripFragment(changeTarget));
                 if (conTarget != null && !conTarget.isEmpty()) {
```

**Comment by raducoravu** [comment]
Looking in the temporary files folder at the ".job.xml", inside it it contains windows-style separators for certain references:

        <file src="file:/D:/projects/eXml/test/dita/EXM-35782/topics/register-license-key.dita"
            uri="topics/register-license-key.dita" path="topics\register-license-key.dita"
            format="dita" skip-chunk="true" target="true" non-conref-target="true"/>
        <file src="file:/D:/projects/eXml/test/dita/EXM-35782/main.ditamap" uri="main.ditamap"
            path="main.ditamap" format="ditamap"/>
        <file src="file:/D:/projects/eXml/test/dita/EXM-35782/glossary/named-user.dita"
            uri="glossary/named-user.dita" path="glossary\named-user.dita" format="dita"
            target="true" non-conref-target="true"/>

so if those URIs are later used in the Java code to compute values for certain targets, this might be the starting point for the problem.

**Comment by jelovirt** [comment]
The `@path` is supposed to be system path, no an URI. So this is correct on Windows.

**Comment by jelovirt** [comment]
@eerohele Is that really the correct location? The `hrefValue` variable is only changed by

```java
hrefValue = hrefValue.substring(fixpath.length());
```

So there's nothing there that could corrupt the directory separators.

**Comment by eerohele** [comment]
No, you're right, I was just looking into it some more and that's not where the problem originates. My earlier proposal fixes the issue, but doesn't address the root cause.

If I do `System.out.println(hrefValue)` at the top of `updateHref`, I can see that the `href` value already uses backslashes at that point, but since I don't know what the input to `TopicRefWriter` is, I don't really know where that `href` value is generated. That's where the problem must be anyway, somewhere in the chunking module, because the `<term>` element doesn't have a `@href` attribute before the chunking phase.

**Comment by jelovirt** [comment]
@eerohele, does

```patch
diff --git a/src/main/java/org/dita/dost/writer/ChunkTopicParser.java b/src/main/java/org/dita/dost/writer/ChunkTopicParser.java
index 9aa62d1..78ad559 100644
--- a/src/main/java/org/dita/dost/writer/ChunkTopicParser.java
+++ b/src/main/java/org/dita/dost/writer/ChunkTopicParser.java
@@ -204,7 +204,7 @@ public final class ChunkTopicParser extends AbstractChunkTopicParser {
                 // TopicRefWriter's updateHref method, very important!!!
                 changeTable.put(path, newpath);
                 // update current element's @href value
-                topicref.setAttribute(ATTRIBUTE_NAME_HREF, getRelativeUnixPath(filePath + UNIX_SEPARATOR + FILE_NAME_STUB_DITAMAP, newpath));
+                topicref.setAttribute(ATTRIBUTE_NAME_HREF, toURI(getRelativeUnixPath(filePath + UNIX_SEPARATOR + FILE_NAME_STUB_DITAMAP, newpath)).toString());
 
                 if (getFragment(parseFilePath) != null) {
                     targetTopicId = getFragment(parseFilePath);

```
Change anything on `hotfix/2.2.3` branch?

**Comment by eerohele** [comment]
@jelovirt: Nope, still getting the same error after applying that patch.

**Comment by raducoravu** [comment]
In AbstractChunkTopicParser there is a call to a deprecated method called "resolveTopic". That seems to return values which contain "\" instead of "/"

**Comment by jelovirt** [comment]
@raducoravu which branch/commit did you use? I've been refactoring that code to use URI for all path processing, replacing File and String usage.

**Comment by raducoravu** [comment]
I think I used the Java code from the 2.2.2 branch.

**Comment by jelovirt** [comment]
@raducoravu are you able to test `develop` code with Windows, to see if this issue still persists?

**Comment by raducoravu** [comment]
I'll test and get back.

**Comment by raducoravu** [comment]
Good news and bad news. With the develop code the transformation does not break anymore.
But the generated XHTML TOC no longer contains the proper titles for the topicrefs and the console contains errors like this:

            [move-meta]   FODC0002: I/O error reported by XML parser processing
            [move-meta]   file:/C:/Users/radu_coravu/Desktop/main/temp/xhtml/Chunk1123907065.dita:
            [move-meta]   C:\Users\radu_coravu\Desktop\main\temp\xhtml\Chunk1123907065.dita (The system cannot find
            [move-meta]   the file specified)


                    [move-meta] file:/C:/Users/radu_coravu/Desktop/main/main.ditamap:6:71: [DOTX027W][WARN]: Unable to retrieve linktext from target: 'Chunk1123907065.dita#register-license-key'. 
                    [move-meta] Recoverable error on line 58 of mappullImpl.xsl:
                    [move-meta]   FODC0002: Document has been marked not available:
                    [move-meta]   file:/C:/Users/radu_coravu/Desktop/main/temp/xhtml/Chunk1123907065.dita
                    [move-meta] file:/C:/Users/radu_coravu/Desktop/main/main.ditamap:7:55: [DOTX061W][WARN]: ID 'unregister-license-key' was used in topicref tag but did not reference a topic element. The href attribute on a topicref element should only reference topic level elements. 
                    [move-meta] file:/C:/Users/radu_coravu/Desktop/main/main.ditamap:7:55: [DOTX023W][WARN]: Unable to retrieve navtitle from target: 'Chunk1123907065.dita#unregister-license-key'. 
                    [move-meta] file:/C:/Users/radu_coravu/Desktop/main/main.ditamap:7:55: [DOTX027W][WARN]: Unable to retrieve linktext from target: 'Chunk1123907065.dita#unregister-license-key'.

**Comment by jelovirt** [comment]
Will try to take a look at this. This new regression bug went past me because we have no Windows CI tests.

**Comment by jelovirt** [comment]
@raducoravu I fixed a bunch of issues I found. Could you retry the `develop` branch on Windows to see if it made any difference?

**Comment by raducoravu** [comment]
I updated, rebuilt and re-run the XHTML output on the sample project, it now breaks:

                dita.inner.topics.html.common:
                     [xslt] Transforming into C:\Users\radu_coravu\Desktop\main\out\xhtml
                     [xslt] Processing C:\Users\radu_coravu\Desktop\main\temp\xhtml\glossary\named-user.dita to C:\Users\radu_coravu\Desktop\main\out\xhtml\glossary\named-user.html
                     [xslt] Loading stylesheet C:\Users\radu_coravu\Documents\dot-develop\src\main\plugins\org.dita.xhtml\xsl\dita2xhtml.xsl
                     [xslt] Processing C:\Users\radu_coravu\Desktop\main\temp\xhtml\topics\register-license-key.dita to C:\Users\radu_coravu\Desktop\main\out\xhtml\topics\register-license-key.html
                     [xslt] C:\Users\radu_coravu\Documents\dot-develop\src\main\xsl\common\functions.xsl:31: Fatal Error! I/O error reported by XML parser processing file:/C:/Users/radu_coravu/Desktop/main/temp/glossary/named-user.dita: C:\Users\radu_coravu\Desktop\main\temp\glossary\named-user.dita (The system cannot find the path specified) Cause: java.io.FileNotFoundException: C:\Users\radu_coravu\Desktop\main\temp\glossary\named-user.dita (The system cannot find the path specified)
                     [xslt] Failed to process null
                
                BUILD FAILED
                C:\Users\radu_coravu\Documents\dot-develop\src\main\build.xml:41: The following error occurred while executing this line:
                C:\Users\radu_coravu\Documents\dot-develop\src\main\plugins\org.dita.xhtml\build_general.xml:168: The following error occurred while executing this line:
                C:\Users\radu_coravu\Documents\dot-develop\src\main\plugins\org.dita.xhtml\build_general.xml:129: Fatal error during transformation using C:\Users\radu_coravu\Documents\dot-develop\src\main\plugins\org.dita.xhtml\xsl\dita2xhtml.xsl: I/O error reported by XML parser processing file:/C:/Users/radu_coravu/Desktop/main/temp/glossary/named-user.dita: C:\Users\radu_coravu\Desktop\main\temp\glossary\named-user.dita (The system cannot find the path specified); SystemID: file:/C:/Users/radu_coravu/Documents/dot-develop/src/main/xsl/common/functions.xsl; Line#: 31; Column#: -1

My temporary files folder is set to "C:\Users\radu_coravu\Desktop\main\temp\xhtml" but the file seems to be searched in: "C:\Users\radu_coravu\Desktop\main\temp\glossary\named-user.dita".

**Comment by raducoravu** [comment]
@jelovirt, does the code on the dev branch work when publishing on Mac OSX the sample project? If so, I could try to investigate the problem more on my side.
About the 2.2.3 branch, either @eerohele can make a small correction of the @href in the "dita-utilities.xsl" (just for the branch) or again a fix in the "AbstractChunkTopicParser" Java class, add some correction of "\" to "/" after the "resolveTopic" method is invoked.

**Comment by jelovirt** [comment]
I've only been using our unit and integration tests, will have to check whether it fails with your test case.

**Comment by jelovirt** [comment]
:+1: for making a simple fix on `hotfix/2.2.3`.

**Comment by jelovirt** [comment]
Could someone try `develop` with 1fc8460 on Windows?

The conversion goes through now, but topicpull still throws errors. Haven't verified, but I think it might be caused by incorrect `path2project-uri` PI values.

**Comment by jelovirt** [comment]
Works for me on OS X. Could someone try on Windows?

**Comment by eerohele** [comment]
Works for me on Windows.

**Comment by jelovirt** [comment]
This has been fixed in 2.3. We could fix this in 2.2.4 we decide to release it and if someone actually fixes this for 2.2.
